{% import 'utils.html' as utils %}

{% extends 'base.html' %}

{% block header %}
<!-- <script src="https://cdn.jsdelivr.net/npm/igv@2.7.4/dist/igv.min.js"></script> -->
<script src="{{ url_for('static', filename='igv.min.js') }}"></script>

<style media="screen">
    .log_div {
        display:block;
        height:500px;
        overflow:scroll;
    }
	/* on hover over a card element */
	.card:hover {
		box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
		transform: scale(1.005);
	}

	/* limit column size to 150 px */
	.table td {
		max-width: 200px;
	}
	
</style>
{% endblock %}

{% block content %}
<div class="mt-4 text-center">

	<h4>TB-Profiler result</h4>
	<div class="">
		Run: {{sample_id}}
	</div>
	<hr>
</div>
{% if result==None %}
    <div class="row justify-content-center">
        <div class="col-md-8">
            {{utils.progress_bar(progress)}}
        </div>
    </div>
    <div class="row justify-content-center text-center">
        <div class="col-md-8">
            Your files are being proessed, please come back later.
        </div>
    </div>

	<div class="row justify-content-center">
		<div class="col-md-8 text-center">
			<div>
				While you wait would you be willing to give some feedback on TB-Profiler? 
				<button data-tf-slider="lS4u9U0O" data-tf-position="right" data-tf-opacity="100" data-tf-iframe-props="title=TBDR feedback" data-tf-transitive-search-params data-tf-medium="snippet" class="btn btn-info p-1 m-2">Click here</button><script src="//embed.typeform.com/next/embed.js"></script>
				to provide feedback.
			</div>
		</div>
	</div>
    <div class="row justify-content-center text-center pt-4">
        <div class="col-mt-8">
            <div class="">

                Want to know whats going on under the hood? <button class="btn btn-outline-dark p-1 m-2" type="button" data-bs-toggle="collapse" data-bs-target="#log" aria-expanded="false" aria-controls="log">Toggle log</button>
            </div>
            <div class="pt-4">

            </div>
        </div>
    </div>
    <div class="row justify-content-center">
        <div class="col-md-8">
            <div class="collapse pt-4" id="log">
                <pre class="log_div"><code>{{log_text}}</code></pre>
            </div>
        </div>
    </div>

	<script>
		window.setTimeout( function() {
			window.location.reload();
		}, 10000);
	</script>

{% else %}

	{% if result["migrated"]==True %}
	<div class="row justify-content-md-center">
        <div class="col-md-10">

					<div class="alert alert-danger" role="alert">
						This result has been migrated from an older version of TB-Profiler. 
						Some information may be missing, and it has not been profiled using 
						the latest mutation database. It is advised to re-run the analysis.
					</div>
		</div>
	</div>
	{% endif %}
	{% if result["qc"]["target_median_depth"] < 30 %}
	<div class="row justify-content-md-center">
        <div class="col-md-10">

					<div class="alert alert-danger" role="alert">
						This average coverage of the target regions is below the recommended 30x.
						Some mutations may not be detected due to low coverage.
					</div>
		</div>
	</div>
	{% endif %}
    <div class="row justify-content-md-center">
        <div class="col-md-10">
    		<div class="card border-dark mb-3">
    			<div class="card-header bg-dark text-white text-center">Summary</div>
    			<div class="card-body text-dark">
    				<p class="card-text"><b>Run ID:</b> {{ result["id"] }}</p>
    				<p class="card-text"><b>Sample name:</b> {{ result["sample_name"] }}</p>
                    <p class="card-text"><b>Date:</b> {{ result["timestamp"] }}</p>
                    <p><b>Number of reads:</b> {{ result["qc"]["num_reads_mapped"]}}</p>
                    <p><b>Percentage reads mapped:</b> {{ result["qc"]["percent_reads_mapped"]}}</p>
                    <p><b>Median coverage:</b> {{ result["qc"]["target_median_depth"]}}</p>
    				<p class="card-text"><b>Strain:</b> {{ result["sub_lineage"] }}</p>
					{% if "spoligotype" in result %}
						<p class="card-text"><b>Spoligotype:</b> {{ result["spoligotype"]["octal"] }}</p>
					{% endif %}
    				<p class="card-text"><b>Drug-resistance:</b> {{ utils.drtype_badge(result["drtype"],result["drtype"]) }}</p>

    				<hr>
					<a class="btn btn-outline-success" href="{{ url_for('static', filename='results/') + result['id']}}.results.csv">Download CSV</a>
					<a class="btn btn-outline-success" href="{{ url_for('static', filename='results/') + result['id']}}.results.txt">Download TXT</a>
					<a class="btn btn-outline-success" href="{{ url_for('static', filename='results/') + result['id']}}.results.json">Download JSON</a>
					
    			</div>
    		</div>
    	</div>
    </div>


	<div class="row justify-content-md-center">
        <div class="col-md-10">
			<div class="card border-dark">
    			<div class="card-header bg-dark text-white text-center">Lineage Table: The lineage is inferred by
    				analysing lineage specific SNPs
    			</div>
    			<div class="card-body">
    				<table class="table">
    					<thead>
    						<tr>
    							<th scope="col">Lineage</th>
    							<th scope="col">Family</th>
    							<th scope="col">RDs</th>
                                <th scope="col">Frequency</th>
    						</tr>
    					</thead>
    					<tbody>
    					{% for row in result["lineage"] %}
    						<tr>
    							<td>{{ row["lineage"] }}</td>
    							<td>{{ row["family"] }}</td>
    							<td>{{ row["rd"] }}</td>
                                <td>{{ row["fraction"]|round(2)|float }}</td>
    						</tr>
    					{% endfor %}
    					</tbody>
    				</table>
    			 </div>
    		</div>
    	</div>
    </div>


    <div class="row justify-content-md-center">
        <div class="col-md-10">

    		<div class="card border-dark">
    			<div class="card-header bg-dark text-white text-center">
                    Drug resistance: This table reports	drug-resistance associated
                    mutations found in known resistance genes
    			</div>
    			 <div class="card-body">
    				<table class="table">
    				  <thead>
    				    <tr>
    				      <th scope="col">Drug</th>
    				      <th scope="col">Resistance</th>
    				      <th scope="col">Supporting mutations</th>
    				      <th scope="col">WHO confidence</th>
    				      <th scope="col">Comment</th>
    				    </tr>
    					</thead>
    				  <tbody>
    					{% for drug_row in result["drug_table"] %}
    						<tr>
								{% if 'drug-rowspan' in drug_row %}
									{% if drug_row['drug-rowspan']>1 %}
    									<td rowspan="2">{{ drug_row["drug"] }}</td>
										{% else %}
										<td>{{ drug_row["drug"] }}</td>
									{% endif %}
								{% endif %}
								{% if 'gene-rowspan' in drug_row %}
									{% if drug_row['gene-rowspan']>1 %}
										<td rowspan="2">{{ drug_row["gene"] }}</td>
									{% else %}
    									<td>{{ drug_row["gene"] }}</td>
									{% endif %}
								{% endif %}

    							<td>{{ drug_row["change"] }}</td>
    							<td>{{ drug_row["confidence"] }}</td>
    							<td>{{ drug_row["comment"] }}</td>
    						</tr>
    					{% endfor %}
    					</tbody>
    				</table>
    			 </div>
    		</div>
    	</div>
    </div>

    

    <div class="row justify-content-md-center">
        <div class="col-md-10">
    		<div class="card border-dark">
    			<div class="card-header bg-dark text-white text-center">
                    Drug resistance-Associated Mutations: This table reports
    				mutations found in candidate resistance genes which have been
    				associated with drug resistance
    			</div>
    		  	<div class="card-body">
    				<table class="table">
    					<thead>
    						<tr>
    							<th scope="col">Gene</th>
    							<th scope="col">Chromosome position</th>
    							<th scope="col">Mutation</th>
                                <th scope="col">Type</th>
                                <th scope="col">Estimated fraction</th>
                                <th scope="col">Drugs</th>
                                <th scope="col">Confidence</th>
                                <th scope="col">Comment</th>
    						</tr>
    					</thead>
    					<tbody>
    					{% for var in result["dr_variants"] %}
							{% for d in var["drugs"] %}
								<tr>
									{{utils.row_val(var["gene_name"],loop.index,var["drugs"]|length) }}
									{{ utils.row_val(var["pos"],loop.index,var["drugs"]|length) }}
									{{ utils.row_val(var["change"],loop.index,var["drugs"]|length,url_for('variants.variant',gene=var['locus_tag'],variant=var['change']))}}
									{{ utils.row_val(var["type"],loop.index,var["drugs"]|length) }}
									{{ utils.row_val(var["freq"]|round(2)|float,loop.index,var["drugs"]|length) }}
									{{ utils.row_val(d['drug']) }}
									{{ utils.row_val(d["confidence"]) }}
									{{ utils.row_val(d["comment"]) }}

								</tr>
							{% endfor %}
    					{% endfor %}
    					</tbody>
    				</table>
    		  	</div>
    		</div>
    	</div>
    </div>

    <div class="row justify-content-md-center">
        <div class="col-md-10">
    		<div class="card border-dark">
    			<div class="card-header bg-dark text-white text-center">Non-Associated Mutations: This table reports
    				mutations found in candidate resistance genes which have not been
    				associated with drug resistance
    			</div>
    		  	<div class="card-body">
    				<table class="table">
    					<thead>
    						<tr>
    							<th scope="col">Gene</th>
    							<th scope="col">Chromosome position</th>
    							<th scope="col">Mutation</th>
                                <th scope="col">Type</th>
    							<th scope="col">
									Estimated fraction
									<i class="bi bi-question-circle" data-bs-toggle="tooltip" data-bs-title="Faction of the reads that support this variant"></i>
								</th>
								<th scope="col">Drugs</th>
								<th scope="col">
									Confidence
									<i class="bi bi-question-circle" data-bs-toggle="tooltip" data-bs-title="Confidence variant/drug association according to the WHO ed. 2 mutation catalogue"></i>
								</th>
    						</tr>
    					</thead>
    					<tbody>
    					{% for var in result["other_variants"] %}
							{% set dlen = var["gene_associated_drugs"]|length %}
							{% for drug in var["gene_associated_drugs"] %}
								<tr>
									{{utils.row_val(var["gene_name"],loop.index,dlen)}}
									{{utils.row_val(var["pos"],loop.index,dlen)}}
									{{utils.row_val(var["change"],loop.index,dlen,url_for('variants.variant',gene=var['locus_tag'],variant=var['change'])) }}
									{{utils.row_val(var["type"],loop.index,dlen)}}
									{{utils.row_val(var["freq"]|round(2)|float,loop.index,dlen)}}
									{{utils.row_val(drug)}}
									{{utils.row_val(var["grading"][drug])}}
								</tr>
								{% endfor %}
    					{% endfor %}
    					</tbody>
    				</table>
    		  	</div>
    		</div>
    	</div>
    </div>

    {% if bam_found %}
        <div class="row justify-content-md-center">
            <div class="col-md-10">
                <div class="card border-dark mb-3">
            		<div class="card-header bg-dark text-white text-center">Pileup</div>
        			<div class="card-body text-dark">
        				<div class="row">
        					<div class="col-md-12" id="pileup-mutations">
        						<div class="">
        							The raw alignment can be visualised here.
        							Click on a mutation to center and zoom the view to
        							the correct genomic location. Drug resistant mutations
        							are coloured in blue and non-associated mutations are
        							coloured in green.
        						</div>
        						<hr>
        						{% for var in result["dr_variants"]%}
        							<button type="button" class="btn btn-outline-primary mb-3" onclick="setPileupView(this)" pos="{{ var['pos'] }}">{{ var["gene_name"] }} {{ var["change"] }}</button>
        						{% endfor %}
        						{% for var in result["other_variants"] %}
        							<button type="button" class="btn btn-outline-success mb-3" onclick="setPileupView(this)" pos="{{ var['pos'] }}">{{ var["gene_name"] }} {{ var["change"] }}</button>
        						{% endfor %}
        					</div>

        				</div>
        				<hr>
            			<div id="igvDiv"></div>
            		</div>
            	</div>
            </div>
        </div>


        <script type="text/javascript">

        	json_result = {{ result|tojson|safe }}
        	variants = {}
        	json_result.dr_variants.forEach((item, i) => {
        		variants[item.pos] = item
        	});
        	json_result.other_variants.forEach((item, i) => {
        		variants[item.pos] = item
        	});
        	console.log(variants)

            document.addEventListener("DOMContentLoaded", function () {
                {% if result['dr_variants']|length > 0 %}
                    locus = "Chromosome:{{result['dr_variants'][0]['pos']|int-10}}-{{result['dr_variants'][0]['pos']|int+10}}"
                {% elif  result['dr_variants']|length > 0 %}
                    locus = "Chromosome:{{result['other_variants'][0]['pos']|int-10}}-{{result['other_variants'][0]['pos']|int+10}}"
                {% else %}
                    locus = "Chromosome:1-100"
                {% endif %}
                var options = {
                    // Example of fully specifying a reference .  We could alternatively use  "genome: 'hg19'"
                    reference: {
                            id: "Mtb",
                            fastaURL: "{{ url_for('static', filename='tbdb.fasta')}}",
                    },
                    locus: locus,
                    tracks: [
                        {
                            type: "alignment",
                            format: "bam",
                            name: "{{ result['id'] }}",
                            url: "{{ url_for('static', filename='results/') + result['id']}}.targets.bam",
                            indexURL: "{{ url_for('static', filename='results/') + result['id']}}.targets.bam.bai",
                        },
                        {
                            type: "annotation",
                            format: "gff3",
                            name: "Genes",
                            url: "{{ url_for('static', filename='tbdb.gff') }}",
                            displayMode: "EXPANDED"
                        }
                    ]
                }
                var igvDiv = document.getElementById("igvDiv");

                igv.createBrowser(igvDiv, options)
                    .then(function (browser) {
                        window.setPileupView = function(objButton){
                            pos = Number(objButton.getAttribute("pos"))
                            browser.search("Chromosome:"+(pos-10)+"-"+(pos+10))
                        }
                    })
            })


        </script>

        <div class="row justify-content-center">
            <div class="col-md-5">
                {{utils.generic_table(result["qc"]["target_qc"],{"target":"Gene","percent_depth_pass":"% pass","median_depth":"Median depth"},"Gene coverage")}}
            </div>
			{% if "spoligotype" in result %}
				<div class="col-md-5">
					{{utils.generic_table(result["spoligotype"]["spacers"],{"name":"Name","seq":"Sequence","count":"Count"},"Spacer count")}}
				</div>
			{% endif %}
        </div>
		<div class="row justify-content-center">
			<div class="col-md-5">
                <div class="card border-dark mb-3">
        			<div class="card-header bg-dark text-white text-center">Software</div>
        			<div class="card-body text-dark">
        				<p><b>TB-Profiler version:</b> {{ result["pipeline"]["software_version"] }}</p>
                        <p><b>Database name:</b> {{ result["pipeline"]["db_version"]["name"] }}</p>
                        <p><b>Database version:</b> {{ result["pipeline"]["db_version"]["commit"] }}</p>


        			</div>
        		</div>
            </div>
		</div>
    {% endif %}
{% endif %}
{% endblock %}
